src/client/Client.cc

```diff
@@ -7999,24 +7999,28 @@ int Client::_readdir_cache_cb(dir_result_t *dirp, add_dirent_cb_t cb, void *p,
   while (true) {
     if (!dirp->inode->is_complete_and_ordered())
       return -EAGAIN;
+ 
+    Dentry *dn = *pd;
+    int r = _getattr(dn->inode, caps, dirp->perms);
+    if (r < 0)
+      return r;
+    
+    //_getattr may change the contents of readdir_cache
     if (pd == dir->readdir_cache.end())
```

Comment:
if dn is deleted, pd is an invalid iterator. invalid iterator != dir->readdir_cache.end()

Reply:
@ukernel 
As long as readdir_cache has the possibility of deleting dentry, there will be this problem. 
The key is that we cannot obtain the address information of the new iterator  after deleting dn.
Any good suggestions, maybe in the process, we don't do _getattr operation?

Reply:
convert pd to vector idx, it's easy to check if a idx is valid.

Reply:
If I judge that pd is illegal, I can only return -EAGAIN, because the iterator after pd is "Iterator invalidation"

Reply:
updated